sys/unix/nleshared.cc

```diff
@@ -0,0 +1,798 @@
+
+#if defined(__linux__) && defined(__x86_64__)
+
+#include <elf.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <dlfcn.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <string_view>
+#include <string>
+#include <cstdio>
+#include <stdexcept>
+#include <system_error>
+#include <vector>
+#include <cstddef>
+#include <cstring>
+#include <unordered_map>
+#include <mutex>
+#include <unordered_set>
+
+namespace nleshared {
+
+struct Fd {
+  int fd = -1;
+  Fd() = default;
+  Fd(int fd) noexcept : fd(fd) {}
+  Fd(const Fd&) = delete;
+  Fd(Fd&& n) noexcept {
+    fd = std::exchange(n.fd, -1);
+  }
+  Fd& operator=(const Fd&) = delete;
+  Fd& operator=(Fd&& n) noexcept {
+    std::swap(fd, n.fd);
+    return *this;
+  }
```

Comment:
I'm almost missing something with `::dup2` here :D

Reply:
> One more question:
> 
> > It will be possible to track memory resources to detect leaks or force cleanup.
> 
> How would you do that? Overriding each of `malloc`, `alloc`, `posix_memalign`, etc?

Yes, exactly.